File: ident-uuid/src/main/scala/net/nicktelford/ident/uuid/rfc4122/DCEUUID.scala
package net.nicktelford.ident.uuid.rfc4122

import net.nicktelford.ident.uuid.UUID

// TODO: implement DCE Security UUID algorithm, detailed here: http://pubs.opengroup.org/onlinepubs/9696989899/chap5.htm#tagcjh_08_02_01_01

object DCEUUID extends RFC4122Factory[DCEUUID] {

  /** Generates the maximum possible UUID for this version. */
  def MaxValue = new DCEUUID(encodeMSB(-1), encodeLSB(-1, -1))

  /** Generates the minimum possible UUID for this version. */
  def MinValue = new DCEUUID(encodeMSB(0), encodeLSB(0, 0))

  /** UUID version this Factory generates. */
  def version = 2
}

/** A Version 2, DCE Security UUID that encodes the POSIX UID/GID */
case class DCEUUID(msb: Long, lsb: Long) extends UUID with RFC4122 {

  val version: Short = 2
}

File: ident-java/src/main/java/net/nicktelford/ident/rfc4122/DCEUUID.java
package net.nicktelford.ident.rfc4122;

/**
 * An RFC4122 Version 2, DCE Security UUID.
 * <p/>
 * This variant is described by The Open Group</a> as part of DCE (Distributed
 * Computing Environment).
 *
 * @see <a href="http://www.ietf.org/rfc/rfc4122">RFC4122</a>
 * @see <a href="http://pubs.opengroup.org/onlinepubs/9696989899/chap5.htm#tagcjh_08_02_01_01">DCE 1.1: Authentication and Security Services</a>
 */
public class DCEUUID extends RFC4122UUID {

    public static final short VERSION = 2;
    private static final Factory FACTORY = new Factory(VERSION);

    public static final DCEUUID MAX_VALUE
            = new DCEUUID(FACTORY.encodeMSB(-1L), FACTORY.encodeLSB(-1L));
    public static final DCEUUID MIN_VALUE
            = new DCEUUID(FACTORY.encodeMSB(0), FACTORY.encodeLSB(0));

    DCEUUID(final long msb, final long lsb) {
        super(msb, lsb);
    }
}
